nucypher/blockchain/eth/agents.py

```diff
@@ -456,16 +469,37 @@ def sample(self, quantity: int, duration: int, additional_ursulas: float = 1.7,
                 f'for {quantity} stakers we need a sample size of at least {n_select}.')
 
         system_random = random.SystemRandom()
-        n_tokens = self.contract.functions.getAllLockedTokens(duration).call()
+        n_tokens, stakers = self.contract.functions.getAllActiveStakers(duration).call()
         if n_tokens == 0:
             raise self.NotEnoughStakers('There are no locked tokens for duration {}.'.format(duration))
 
         for _ in range(attempts):
             points = sorted(system_random.randrange(n_tokens) for _ in range(n_select))
             self.log.debug(f"Sampling {n_select} stakers with random points: {points}")
 
-            addresses = set(self.contract.functions.sample(points, duration).call())
-            addresses.discard(str(BlockchainInterface.NULL_ADDRESS))
+            addresses = set()
+
+            previous_point = 0
```

Comment:
Just curious, how important the performance of this algorithm is, and whether the choice of linear search was intentional, or just the simplest one? As far as I understand it, the situation is as follows:

Let `M = stakers_len` and `N = n_select`. Then for different weighted selection algorithms we will have:
* linear search: `O(N log N) + O(M)`
* binary search: `O(M) + O(N log M)`
* hopscotch: `O(M log M) + O(N log M)` (but the second term has a much smaller constant factor than that in binary search, so it often turns out much faster)
In the application, `M` is probably large, and `N` is small. So if `getAllActiveStakers()` can be cached, the last two approaches will be faster. Although it is a smart contract call, so I'm guessing it can't?

Reply:
Good points here!
I think, `getAllActiveStakers` can be cached for 1 day (for the current period) since this list doesn't change during this time. And it's a time-consuming call too! After it's done, optmizing weighted selection algorithm may help further

Reply:
While there's room for optimization here, I have some comments:
1) `getAllActiveStakers(_periods)` takes a `_periods` parameter for the duration of the policy, so that makes cache difficult and not that useful, specially if it only makes sense for that day.
2) Cached where? Caching in the smart contract costs gas, while current approach is free since it's a call.
3) Binary search makes a lot of sense, actually. Current approach was linear since it's a direct translation of the original sampling method in the smart contract, where sorting the stakers array is not convenient. But it does make sense to just traverse the array of random points, and for each point, perform binary search on the stakers array. However, we'd have to modify `getAllActiveStakers(_periods)` to return a sorted list of stakers.
4) Where can I find a description of hopscotch search? Is it related to [hopscotch hashing](https://en.wikipedia.org/wiki/Hopscotch_hashing)?

Reply:
4. I read about it here: https://blog.bruce-hill.com/a-faster-weighted-random-choice

Reply:
2. Yeah, indeed, caching doesn't work well here since the policy duration could be anything. Only if we have many policies of the same duration